Intercept «incompatible type in overriding» errors

Example:
```scala
trait Base {
  type Env
  def makeEnv: ZLayer[Any, Nothing, Env]
}

object ZMain extends Base {
  override type Env = String

  override def makeEnv: ZLayer[Any with Long, Nothing, String] = ZLayer.succeed("42")
}
```

Output:
```
billing/billing/billing-tms/app/src/ZMain.scala:34: error: incompatible type in overriding
def makeEnv: zio.ZLayer[Any,Nothing,ru.yandex.vertis.billing.tms.ZMain0.Env] (defined in trait Base);
 found   : zio.ZLayer[Any with Long,Nothing,String]
 required: zio.ZLayer[Any,Nothing,ru.yandex.vertis.billing.tms.ZMain0.Env]
    (which expands to)  zio.ZLayer[Any,Nothing,String]
  override def makeEnv: ZLayer[Any with Long, Nothing, String] = ZLayer.succeed("42")
               ^
zio.ZLayer[Any with Long,Nothing,String] <: zio.ZLayer[Any,Nothing,ru.yandex.vertis.billing.tms.ZMain0.Env]?
```

Looks like magic regex ignores such errors because it is `incompatible type in overriding`, not a `type mismatch` 🤔

--- a/src/main/scala-2/clippy/ZIOErrorReporter.scala
+++ b/src/main/scala-2/clippy/ZIOErrorReporter.scala
@@ -1,7 +1,7 @@
 package clippy
 
 import clippy.ansi.AnsiStringOps
-import clippy.utils.{Info, IsZIOTypeMismatch}
+import clippy.utils.{Info, IsZIOTypeError, ErrorKind}
 
 import scala.reflect.internal.Reporter
 import scala.reflect.internal.util.Position
@@ -10,12 +10,32 @@ import scala.tools.nsc.reporters.FilteringReporter
 
 final class ZIOErrorReporter(val settings: Settings, underlying: Reporter, showOriginalError: Boolean)
     extends FilteringReporter {
-  private def makeError(found: Info, required: Info, msg: String): String = {
+
+  private case class TitleMessages(envMismatch: String, errorMismatch: String, returnMismatch: String)
+  private val mismatchMessages = TitleMessages(
+    envMismatch = "Your effect requires the following environment, but it was not provided:",
+    errorMismatch = "Your effect has an error type mismatch:",
+    returnMismatch = "Your effect has a return type mismatch:"
+  )
+
+  private val overridingMessages = TitleMessages(
+    envMismatch =
+      "Your effect requires the following environment in overriding, but it was not presented in original type:",
+    errorMismatch = "Your effect has an error type mismatch in overriding:",
+    returnMismatch = "Your effect has a return type mismatch:"
+  )
+
+  private def makeError(found: Info, required: Info, msg: String, errorKind: ErrorKind): String = {
+    val titleMessages = errorKind match {
+      case ErrorKind.Overriding   => overridingMessages
+      case ErrorKind.TypeMismatch => mismatchMessages
+    }
+
     def envMismatch = {
       val diff = found.R -- required.R -- Set("Any")
 
       if (diff.nonEmpty) {
-        Some(s"""Your effect requires the following environment, but it was not provided:
+        Some(s"""${titleMessages.envMismatch}
                 |
                 |${diff.map(r => s"${"❯ ".red}${r.bold}").toList.mkString("\n")}
                 |""".stripMargin)
@@ -26,7 +46,7 @@ final class ZIOErrorReporter(val settings: Settings, underlying: Reporter, showO
       val diff = found.E != required.E && required.E != "Any"
 
       Option.when(diff) {
-        s"""Your effect has an error type mismatch:
+        s"""${titleMessages.errorMismatch}
            |
            |${"❯ ".red}${"Required"}: ${required.E.bold}
            |${"❯ ".red}${"Found   "}: ${found.E.bold}
@@ -38,7 +58,7 @@ final class ZIOErrorReporter(val settings: Settings, underlying: Reporter, showO
       val diff = found.A != required.A
 
       Option.when(diff) {
-        s"""Your effect has a return type mismatch:
+        s"""${titleMessages.returnMismatch}
            |
            |${"❯ ".red}${"Required"}: ${required.A.bold}
            |${"❯ ".red}${"Found   "}: ${found.A.bold}
@@ -64,8 +84,8 @@ final class ZIOErrorReporter(val settings: Settings, underlying: Reporter, showO
 
   override def doReport(pos: Position, msg: String, severity: Severity): Unit =
     (severity, msg) match {
-      case (Reporter.ERROR, IsZIOTypeMismatch(found, required)) =>
-        val error = makeError(found, required, msg)
+      case (Reporter.ERROR, IsZIOTypeError(kind, found, required)) =>
+        val error = makeError(found, required, msg, kind)
         underlying.error(pos, error)
       case _ =>
         severity match {

--- a/src/main/scala/clippy/utils.scala
+++ b/src/main/scala/clippy/utils.scala
@@ -5,6 +5,12 @@ import scala.util.matching.Regex
 object utils {
   case class Info private (R: Set[String], E: String, A: String)
 
+  sealed trait ErrorKind
+  object ErrorKind {
+    case object Overriding   extends ErrorKind
+    case object TypeMismatch extends ErrorKind
+  }
+
   object Info {
     // The original error message will needlessly dealias nested definitions.
     private val substitutions = Map(
@@ -34,16 +40,20 @@ object utils {
       }
   }
 
-  object IsZIOTypeMismatch {
+  object IsZIOTypeError {
     val mismatch = raw"zio\.(ZIO|ZLayer)\[(.+),([^,\]]+),(.+)\]".r
-    def unapply(msg: String): Option[(Info, Info)] =
-      if (msg.contains("type mismatch;")) {
-        mismatch.findAllMatchIn(msg).toList match {
-          case _ :: Info(found) :: _ :: Info(required) :: Nil => Some(found, required)
-          case Info(found) :: Info(required) :: _             => Some(found, required)
-          case _                                              => None
-        }
-      } else None
+
+    private def findMismatches(msg: String, kind: ErrorKind): Option[(ErrorKind, Info, Info)] =
+      mismatch.findAllMatchIn(msg).toList match {
+        case _ :: Info(found) :: _ :: Info(required) :: Nil => Some(kind, found, required)
+        case Info(found) :: Info(required) :: _             => Some(kind, found, required)
+        case _                                              => None
+      }
+
+    def unapply(msg: String): Option[(ErrorKind, Info, Info)] =
+      if (msg.contains("type mismatch;")) findMismatches(msg, ErrorKind.TypeMismatch)
+      else if (msg.contains("incompatible type in overriding")) findMismatches(msg, ErrorKind.Overriding)
+      else None
   }
 
   object IsCannotProveMismatch {

